### DISCUSSION TITLE
add_cron returns "tasmota.add_cron("*/5 * * * * *", refresh_display(1), "every_5_s")"

### DISCUSSION THREAD

**Discussion by indigmas** [arendst/Tasmota#20996] [Q&A] [answered]
I want to show my sonsors data for 1 second , every 5 seconds and use this code:
```
def refresh_display(wait_sec)
    import json
    import string
    var s_map=json.load(tasmota.read_sensors())
    tasmota.cmd("DisplayText [zl2c1s3]" +"T:"+ str(s_map['DHT11']['Temperature']))
    tasmota.cmd("DisplayText [l1c1s3]" +"H:"+ str(s_map['DHT11']['Humidity'])+"%")   
    tasmota.delay(wait_sec*1000)
    tasmota.cmd("DisplayText [z]")
end
```
and
`tasmota.add_cron("*/5 * * * * *", refresh_display(1), "every_5_s")`
what is wrong, because it brings me that error:
```
BRY: Exception> 'type_error' - BRY: argument must be a function
stack traceback:
	<unknown source>: in function `check_not_method`
	<unknown source>: in function `add_cron`
	<unknown source>: in function `main`
```

**Comment by s-hadinger** [comment, marked as answer]
`add_cron` needs a function to call, but `refresh_display(1)` is just a call that returns a value. It's not a function.

You need a closure for this:
`tasmota.add_cron("*/5 * * * * *", def () refresh_display(1) end, "every_5_s")`

or a simpler syntax:
`tasmota.add_cron("*/5 * * * * *", / -> refresh_display(1), "every_5_s")`

**Reply by sfromis** [reply]
If the called function can tolerate having no argument, you can also do like this:
`tasmota.add_cron("*/5 * * * * *", refresh_display, "every_5_s")`
This is still passing a reference to the function variable, without calling it right away.

**Reply by indigmas** [reply]
ok, that works fine :) - thank you very much.
for the future:
- what is the difference between call and function? (function returns a value?)
- how is a function defined?

**Reply by sfromis** [reply]
The two examples provided by s-hadinger is defining an anonymous function in place, which is then calling `refresh_display` function you already defined. When you define a named function, you are effectively creating a variable named after the function, and with the content being the code for the function. The 2nd example is abbreviated "syntactic sugar", which for simple cases results in exactly the same function as the 1st. Such inline functions can get really useful as closures when you get a bit more advanced.

To call the function contained in the variable, you add an argument list in `()`. When not having an argument list, you are passing the variable value (the function). And using `add_cron` requires a function, which can be such a function reference via a variable (without arguments). Or you can use an anonymous function definition directly where you could also have just a function reference.

Here's an illustrative example: `(def (name) print('hello ' + name) end)('world')`
First an anonymous function is defined within the first set of `()` returning the function content. Instead of putting that into a variable, it is called directly via the last set of `()` and an argument.
